usersession/userd/launcher.go

```diff
@@ -123,6 +130,157 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface. The desktop_file_id is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	desktop_file, err := desktopFileIdToFilename(existsOnFileSystem, desktop_file_id)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktop_file)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Before rejoining args to create a new command-line: escape "\" escapes; escape "\"" quotes; and, wrap each arg in "\"" quotes
+	for i := 0; i != len(args); i++ {
+		args[i] = strings.Replace(args[i], "\\", "\\\\", -1)
+		args[i] = strings.Replace(args[i], "\"", "\\\"", -1)
+		args[i] = "\"" + args[i] + "\""
+	}
+
+	// Invoke indirectly via sh to unparent and avoid potentially leaving a zombie
+	cmd := exec.Command("sh", "-c", strings.Join(args, " ")+"&")
```

Comment:
I'm not sure we should be doing this via sh. Would this work?
```
cmd := exec.Command(args)
cmd.SysProcAttr = &syscall.SysProcAttr{
    Setpgid: true,
}
... <setup env>
if err := cmd.Start(); err != nil {
    return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
}
```
Then we don't need the shell quoting bits above.

Reply:
No, that still leads to zombies...
```plain
$ ps -eaf | grep kodi
alan      192927  192920  0 10:21 ?        00:00:00 [kodi] <defunct>
```

Reply:
I'm not too happy about needing to go through sh to get that intermediate child/reparenting going (also `sh` could be bash/dash/whatever so it's unclear what job control will do). OTOH there seems to be no standard way to start the app from a desktop file using gio/gvfs/gtk/kio helpers either. If only xdg-open did the right thing here and launched the app, but of course it doesn't. Then `userd` is a long lived process so it can't really exist and let the child be reparented.

Maybe we could have a mode of userd in which it just launches the `*.desktop` file app? Then the current userd could call `/proc/self/exe launch-desktop foo.desktop`. That new process calls `cmd.Start()` on the app, does not wait and exits. The parent waits for the desktop launcher to return. In theory the grandchild should then end up under the subreaper, gnome-session or systemd --user, would it not?

I'd be interesting to hear from @zyga, @mvo5 or @jdstrand on this.

Reply:
Yup. It is very possibly my ignorance of Go, but I couldn't find a way to double fork and exec.

Reply:
@bboozzoo - I meant to say that yesterday I ran out of time to look at this part, but I think we're seeing this due to go's lack of fork(), so we can't do the standard double fork() technique to daemonize the process and have it reparent in the way we'd like (this is what `sh -c ... &` is working around). I did try setsid using `Setsid: true,` but that didn't do it. Calling wait() in a go routine (ie `go cmd.Wait()`) after `cmd.Start()` does though:
```
cmd := exec.Command(...
... env ...
if err := cmd.Start(); err != nil {
        return dbus.MakeFailedError(err)
}

// XXX: this avoids defunct processes but causes userd to persist
// until all children are gone (currently, this is not a problem since
// userd is long running once started)
go cmd.Wait()
return nil
```
That is better than cmd.Run() which makes the caller hang and better than not calling wait since that ends up with defunct processes until userd exits. I believe it means that userd can't time out until all the children die (I didn't look at this extensively), but AFAICS, once userd starts it keeps running (I have a userd on my system today that has been running for 6 days). I therefore think this is *far* more acceptable than calling `sh -c ... &` and suggest that this code be changed to that with the something along the lines of the comment I gave (cc @AlanGriffiths).

To solve that, perhaps there is something we can do in the tomb code to kill itself if all we are doing is waiting on children. We could also consider a small C daemonizer. There may be other options like looking into https://golang.org/pkg/syscall/#ForkExec, et al.